src/auth/auth.service.ts

```diff
@@ -1,27 +1,85 @@
-import { Injectable, BadRequestException,NotAcceptableException } from '@nestjs/common';
+import { Injectable, BadRequestException } from '@nestjs/common';
 import { UserSignupDto } from './dto/user-signup.dto';
 import { InjectRepository } from '@nestjs/typeorm';
 import { User } from '../entities/user.entity';
 import { Repository } from 'typeorm';
 import { ZaLaResponse } from 'src/common/helpers/response';
+import { UserSigninDto } from './dto/user-signin.dto';
+import { JwtHelperService } from './jwtHelper.service';
 
 @Injectable()
 export class AuthService {
   constructor(
     @InjectRepository(User)
     private userRepo: Repository<User>,
-    
-  ){}
+    private jwtHelperService: JwtHelperService,
+  ) {}
   async signup(user: UserSignupDto) {
     const userdata = Object.assign(new User(), user);
     const newUser = await this.userRepo.save(userdata).catch(async (e) => {
       throw new BadRequestException(
-        ZaLaResponse.BadRequest(
-          'Duplicate Values',
-          'The Email already exists',
-        ),
-        );
+        ZaLaResponse.BadRequest('Duplicate Values', 'The Email already exists'),
+      );
     });
-    return newUser
+    return newUser;
+  }
+
+  async signin(
+    dto: UserSigninDto,
+    values: { userAgent: string; ipAddress: string },
+  ) {
+    const user = await this.userRepo.findOne({ where: { email: dto.email } });
+
+    if (!user)
+      throw new BadRequestException(
+        ZaLaResponse.BadRequest('Not found', 'Invalid Credentials!'),
+      );
+
+    const hash = await this.jwtHelperService.hashPassword(
+      dto.password,
+      user.password.split(':')[0],
+    );
+
+    let isPasswordCorrect = hash == user.password;
+    if (!isPasswordCorrect)
+      throw new BadRequestException(
+        ZaLaResponse.BadRequest('Access Denied!', 'Incorrect Credentials'),
+      );
+
+    const tokens = await this.getNewRefreshAndAccessTokens(values, user);
+
+    // add userInfo to the list of user history and return along in the response
+    await this.userRepo.update(user.id, {
+      history: [...user.history, dto.userInfo],
```

Comment:
Line 53, you should be passing in the values argument, not dto.userInfo

Reply:
The dto.userInfo contains the object and its values

Reply:
Front end cant be passing browser info, you have to pick it from the req object, which makes the dto unnecessary

Reply:
@johnayinde  If you tested this code you'll realise that the req data you are getting from the controller, which is the same data you called values in the service is not being saved.